x/concentrated-liquidity/types/keys.go

```diff
@@ -113,10 +116,30 @@ func keyTickPrefixByPoolIdPrealloc(poolId uint64, preAllocBytes int) []byte {
 	return key
 }
 
-// KeyFullPosition uses pool Id, owner, lower tick, upper tick, joinTime and freezeDuration for keys
-func KeyFullPosition(poolId uint64, addr sdk.AccAddress, lowerTick, upperTick int64, joinTime time.Time, freezeDuration time.Duration) []byte {
+// KeyFullPosition uses pool Id, owner, lower tick, upper tick, joinTime, freezeDuration, and positionId for keys
+func KeyFullPosition(poolId uint64, addr sdk.AccAddress, lowerTick, upperTick int64, joinTime time.Time, freezeDuration time.Duration, positionId uint64) []byte {
 	joinTimeKey := osmoutils.FormatTimeString(joinTime)
-	return []byte(fmt.Sprintf("%s%s%x%s%d%s%d%s%d%s%s%s%d", PositionPrefix, KeySeparator, addr.Bytes(), KeySeparator, poolId, KeySeparator, lowerTick, KeySeparator, upperTick, KeySeparator, joinTimeKey, KeySeparator, uint64(freezeDuration)))
+
+	var builder strings.Builder
+	builder.Grow(len(PositionPrefix) + 9*len(KeySeparator) + len(addr.Bytes())*2 + 5*20) // pre-allocating the buffer
+
+	builder.Write(PositionPrefix)
+	builder.WriteString(KeySeparator)
+	builder.WriteString(hex.EncodeToString(addr.Bytes()))
+	builder.WriteString(KeySeparator)
+	builder.WriteString(strconv.FormatUint(poolId, 10))
```

Comment:
This is how we can serialize integers in a space-efficient way per my earlier `bytes.Buffer` suggestion:

```go
	// Create a buffer to store the bytes
	var buf bytes.Buffer

	// Serialize the uint64 value using PutUvarint
	n := binary.PutUvarint(buf.Bytes()[:binary.MaxVarintLen64], value)

	// Truncate the buffer to the actual number of bytes used by the encoded value
	buf.Truncate(n)
```